src/serialization.jl

```diff
@@ -21,36 +31,113 @@ mutable struct TrialPreV006
-                      p.gcsample, p.time_tolerance, p.memory_tolerance)
+function recover(x::Vector)
+    length(x) == 2 || throw(ArgumentError("Expecting a vector of length 2"))
+    typename = x[1]::String
+    fields = x[2]::Dict
+    T = eval(parse(typename))::Type
+    fc = fieldcount(T)
+    xs = Vector{Any}(fc)
+    for i = 1:fc
+        ft = fieldtype(T, i)
+        fn = String(fieldname(T, i))
+        xs[i] = if ft in supported_types
+            recover(fields[fn])
+        else
+            convert(ft, fields[fn])
+        end
+    end
+    T(xs...)
 end
 
-function JLD.readas(t::TrialPreV006)
-    new_times = convert(Vector{Float64}, t.times)
-    new_gctimes = convert(Vector{Float64}, t.gctimes)
-    return Trial(t.params, new_times, new_gctimes, t.memory, t.allocs)
+function save(filename::AbstractString, args...)
+    isempty(args) && throw(ArgumentError("Nothing to save"))
+    if !endswith(filename, ".json")
+        noext, ext = splitext(filename)
+        msg = if ext == ".jld"
+            "JLD serialization is no longer supported. Benchmarks should now be saved\n" *
+            "in JSON format using `save(\"$noext\".json, args...)`. You will need to\n" *
+            "convert existing saved benchmarks to JSON in order to deserialize them with\n" *
+            "this version of BenchmarkTools."
+        else
+            "Only JSON serialization is supported. Use `save(\"$noext.json\", args...)`."
+        end
+        throw(ArgumentError(msg))
+    end
+    goodargs = Any[]
+    for arg in args
+        if arg isa String
+            warn("Naming variables in serialization is no longer supported.\nThe name " *
+                 "will be ignored and the object will be serialized in the order it appears " *
+                 "in the input.")
+            continue
+        elseif !any(T->arg isa T, supported_types)
+            throw(ArgumentError("Only BenchmarkTools types can be serialized."))
+        end
+        push!(goodargs, arg)
+    end
+    isempty(goodargs) && error("Nothing to save")
+    open(filename, "w") do io
+        JSON.print(io, [VERSIONS, goodargs])
+    end
 end
 
-function save(filename, args...)
-    JLD.save(filename, VERSION_KEY, VERSIONS, args...)
-    JLD.jldopen(filename, "r+") do io
-        JLD.addrequire(io, BenchmarkTools)
+function load(filename::AbstractString, args...)
+    if !endswith(filename, ".json")
+        noext, ext = splitext(filename)
+        msg = if ext == ".jld"
+            "JLD deserialization is no longer supported. Benchmarks should now be saved\n" *
+            "in JSON format using `save(\"$noext\".json, args...)`. You will need to\n" *
+            "convert existing saved benchmarks to JSON in order to deserialize them with\n" *
+            "this version of BenchmarkTools."
+        else
+            "Only JSON deserialization is supported. Use `load(\"$noext.json\", args...)`."
+        end
+        throw(ArgumentError(msg))
+    end
```

Comment:
Might want to break out this duplicated assert into its own function (either rewriting the message to be the same for load/save or just pass a flag giving the context).